design/include-cleaner.md

```diff
@@ -0,0 +1,167 @@
+in the main file. When you open a file, clangd will analyze the symbols the
+file uses and mark all headers defining these symbools as "used". The warnings
+will be issues for the headers that are included but not marked as "used".
+Example:
+
+```c++
+// foo.h
+class Foo {};
+```
+
+```c++
+// bar.h
+class Bar {};
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+#include "bar.h" // <- Will be marked as unused and suggested to be removed.
+
+int main() {
+  Foo f;
+}
+```
+
+Here, `main.cpp` only makes use of symbols from `foo.h` and removing `#include
+"bar.h"` prevents unnecessary parsing of `bar.h` and allows breaking the
+dependncy on it.
+
+### Deciding what headers are used
+
+Clangd relies on Clang AST to deduce which headers are used and which aren't,
+the whole Include Cleaner decision process is described below.
+
+#### Scanning the main file
+
+First, Include Cleaner will build the AST for the main file (the file currently
+opened in the editor). After the AST is built, Include Cleaner will recursively
+visit the AST nodes and collect locations of all referenced symbols (e.g.
+types, functions, global variables). The AST will contain macro and `auto`
+expansions. For example, in this case, `foo.h` will be considered as used:
+
+```c++
+// foo.h
+
+int foo() { return 42; }
+```
+
+```c++
+// macro.h
+
+#include "foo.h"
+
+#define FOO foo
+```
+
+```c++
+// main.cpp
+
+#include "macro.h"
+
+int main() {
+  FOO();
+}
+```
+
+This means that Include Cleaner is conservatively treating symbols in the
+expanded code as usages as opposed to only explicitly spelled symbols.
+
+Include Cleaner will also traverse the unexpanded macro tokens to see where
+their definitions come from.
+
+After the process is complete, the declaration and definition locations will be
+collected and passed to the next stage.
+
+#### Marking the headers as used
+
+`SourceLocation` instances collected at the previous step will be converted to
+`FileID`s and deduplicated. In this stage, it is important to attribute the
+locations in some headers to their includes. Some of the `FileID`s correspond
+to non self-contained headers, meaning the user should actually include their
+parent rather than a header itself. Other important headers with this property
+are the ones manually marked as private through `IWYU pragma: private`. For
+them, the user explicitly asks includers to consider the public header.
+
+After the responsible headers are collected, they only step left is producing
+diagnostics for unused headers.
+
+#### Issuing warnings
+
+After most of the work is done, Include Cleaner needs to decide which headers
+are not used in the main file. All inclusions are scanned and checked for
+containing `IWYU pragma: keep`; if they do not, they are not used and will be
+warned about.
+
+### IWYU pragmas
+
+IWYU tool offers a set of
+[pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md),
+the most notable of which are:
```

Comment:
rather than saying `the most notable` i think we should say `Include Cleaner respects the following:`